### PR TITLE
CI build script improvements, #1918

### LINKED ISSUE #1918: Speed up the build timings of CircleCI jobs
Jobs running on CircleCI: 
- build_debug
- test_unit 
- test_instrumented (only on master branch
- report_coverage   (only on master branch)

There is still scope to speed up the time taken to run each job. Some of the ideas to try out are:
- Try out different parameters available in `environment` , `_JAVA_OPTIONS` in `config.yml` file.
- Cache the android dependencies for `pmd`, `findbugs` so that they are not downloaded during every build.
- Optimize the  gradle tasks (for eg. check if the dependency `testDebugUnitTest` is required in `jacocoTestReport`, if not then remove it)
- Explore other ideas. Read this [link](https://circleci.com/blog/why-is-my-build-slow/) as well and many others have posted blogs for the same as well

### PR DESCRIPTION
**Pull request by andreimarcut**
Closes #1918 

Enabled 3rd party dependency caching for debug_build and unit_test jobs.
These changes should improve the CI build process time. 

#### What has been done to verify that this works as intended?
There are no functional code changes. These changes only impact the build process. 

#### Why is this the best possible solution? Were any other approaches considered?
There are other possible additional improvements that I have in mind, but they do not exclude the improvement gains from this partial contribution. Otherwise said, they would stack with it. I am about to start a topic on how things could be further improved moving forward.

#### Are there any risks to merging this code? If so, what are they?
I couldn't find particular CI scripts coding standards, or related preferences. But if there are any, they could be impacted.

#### Do we need any specific form for testing your changes? If so, please attach one.
I think a "smoke test" is enough.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/opendatakit/docs/issues/new) and include the link below.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lint findbugs` and confirmed all checks still pass.
- [ ] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml]: These can probably be removed since you decided not to use them.

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml]: It'd be nice to have a comment inline saying you used this to avoid an update.

**Review by yanokwa — COMMENTED** [review]
This is fantastic, @andreimarcut. Thanks for sending this in. Do you have a sense of what kind of speed up we are looking at?

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml]: I did comment on commit #840562 message: a select android api 27 machine image to avoid an unnecessary platform update

Is an in-line comment desired as well?

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml]: I did comment on commit #840562 message: a select android api 27 machine image to avoid an unnecessary platform update

Is an in-line comment desired as well?

Comment (yanokwa) [reply .circleci/config.yml]: Yes, I saw that comment, but it's the kind of thing that might be confusing in the future and it's nice not to go back in history. I don't feel strongly about it though.

Comment (andreimarcut) [reply .circleci/config.yml]: Ok, I'll add a comment then

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml]: Indeed, there's no use in the process. I only found it convenient to leave it there, for easy toggling in future experiments

File: .circleci/config.yml
Comment (shobhitagarwal1612) [inline .circleci/config.yml]: Since the regex allows all branches so nothing is filtered out. So is the `filters` section required?

File: .circleci/config.yml
Comment (shobhitagarwal1612) [inline .circleci/config.yml]: Same as above

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml]: It's not required, I'll remove it

**Comment by andreimarcut** [comment]
@yanokwa @shobhitagarwal1612 I've updated the PR with extra improvements, gained another ~1min roughly in build_debug job and stabilized the random gradle crashes by limiting gradle max heap memory to optimal level

**Review by shobhitagarwal1612 — APPROVED** [review]

**Comment by andreimarcut** [comment]
The 'random gradle crash' I'm talking about is on the CI machine:

> * What went wrong:
> Execution failed for task ':collect_app:findbugs'.
> > Failed to run Gradle FindBugs Worker
>    > Process 'Gradle FindBugs Worker 1' finished with non-zero exit value 137
> 

I also experienced it randomly during my tests, before tweaking the memory settings in the CI scripts (some times it failed, other times it worked - without any code changes). Afterwards, it didn't occur any more.

Specifically, I noticed that the gradle processes running during the build already consume the machine's available memory, and then, when the findbugs task starts, it actually spawns a new process, which fails to bootstrap - most likely due to lack of memory resources
I believe this issue should be further analyzed and monitored from the historic CI build stats page as well as in the future

Also encountered on these PRs:
https://circleci.com/gh/shobhitagarwal1612/collect/987
https://circleci.com/gh/opendatakit/collect/3009

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml:123]: It feels like 3x speed up is a worthy tradeoff, but, I have to ask. How likely is it that there will be an issue findable by lint that is in the release build that is not in the debug build?

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml]: You have a double period at the end.

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml]: Drop the double period.

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml:228]: What are the trade offs. If it's something to try empirically, then might be easier to file an issue for a contributor to try.

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml:236]: What are the trade offs. If it's something to try empirically, then might be easier to file an issue for a contributor to try.

File: .circleci/config.yml
Comment (yanokwa) [inline .circleci/config.yml]: A comment here about how you reached these magic flags would be nice.

**Review by yanokwa — CHANGES_REQUESTED** [review]
@andreimarcut I'm really excited about this PR because it reduces build time and that means increased productivity for all contributors!

I think it's ready to merge, but I'd like to wait until after the next Collect release (in the next 24 hours). Given we have some time, I figured I'd leave some nitpicky feedback...

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml:123]: Good point. Indeed we could miss issues on what was not analyzed/checked. 
But actually, the flavor difference itself is just a fraction of the project (in fact it's just the build configuration).
On the other hand, the debug build shouldn't be concerned with other configurations and the way it used to be before this change, every debug was burdened (although out of scope).

There are two approaches I'd weight-balance here:
1. run lintRelease or lintOdkCollectRelease on every debug build (and drop lintDebug entirely from the debug build)
2. run lintDebug on debug builds and run lintRelease or lintOdkCollectRelease on release builds

With (1) during debug build, we'd know any release-flavor lint violations early, but we may miss debug version violations. I don't like this approach because it's a compromise.
With (2) the debug build deals with it's own quality and the release build is separately concerned with it's own quality

My suggestions is to go with (2) and keep the concern within it's scope. I'd rather re-run the quality checks on release build to double-check them, rather than run extra release checks on every debug build.

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml]: Will drop

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml]: Will drop

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml:236]: It's mostly an organizational policy than a technical decision. There's no direct dependency between these two jobs. However, it's likely that in case one fails, the other would fail too. So one matter here is wasting (cloud) resources when not needed and on the other side the counterweight is if we care about possibly wasting these resources on failed tests more than we care about finishing the builds quicker in the event of a success?

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml:228]: Two lines below the comment, there's another comment stating 
"         Although the unit testing task could run in parallel with build task,
         in such case, it would not benefit from the previous caching in the event of
         any change in the .gradle script files."

For me that is enough to choose that unit tests depend on the build task. However, in case the other way is preferred, based on things I don't know yet i.e. organization policy (I've just jumped straight here out of nowhere ;), it's a trivial task to toggle. So, please advise.

File: .circleci/config.yml
Comment (andreimarcut) [reply .circleci/config.yml]: Here it is, I'll add it as comment in the source too:
       Runtime configuration build tests
       ./gradlew clean assembleDebug
       Xms128 - the default on most systems is 256m but gradle spawns some java subprocesses during build that don't need 256m
       Xmx512m 1m30s
       Xmx768m 1m17s
       Xmx1024m ~1min
       Xmx1536m ~1min
       -XX:+HeapDumpOnOutOfMemoryError in case of an out of memory the java process will dump a heap file on the file system.
       The heap file can be analyzed to figure out the root cause or the memory leak.
       -DpreDexEnable=false disables pre-dexing step. Pre-dexing is good for local incremental builds during development,
       But on a build machine, the performance impact is not worthy because on a CI, every build should be clean.

File: .circleci/config.yml
Comment (yanokwa) [reply .circleci/config.yml:228]: I think it's almost always better that we get faster builds, so you made the right choice. Not bad from jumping straight out of nowhere!

File: .circleci/config.yml
Comment (yanokwa) [reply .circleci/config.yml:236]: Finishing faster at all costs. We currently don't have to worry about waste.

File: .circleci/config.yml
Comment (yanokwa) [reply .circleci/config.yml:123]: Agreed with this approach! I'll note this in our release process. cc @lognaturel

**Review by yanokwa — APPROVED** [review]

**Comment by yanokwa** [comment]
@andreimarcut I'm going to merge this now because it's an improvement with very little side-effects. If you can, please send in another PR with an updated comment regarding https://github.com/opendatakit/collect/pull/2247#discussion_r191402107 and any tweaks to finish faster in the event of success (https://github.com/opendatakit/collect/pull/2247#discussion_r191398612)

**Comment by yanokwa** [comment]
Also, thanks again for sending in this PR!

**Pull request merged**
